storage/innobase/row/row0ins.cc

```diff
@@ -1867,6 +1870,24 @@ row_ins_check_foreign_constraint(
 			&check_table->n_foreign_key_checks_running, 1);
 
 		lock_wait_suspend_thread(thr);
+#ifdef WITH_WSREP
+		trx_mutex_enter(trx);
+		switch (trx->error_state) {
+		case DB_DEADLOCK:
+			if (wsrep_debug) {
+				ib::info() <<
+				"WSREP: innodb trx state changed during wait "
+				<< " trx: " << trx->id << " with error_state: "
+				<< trx->error_state << " err: " << err;
+			}
+			err = trx->error_state;
+			break;
+		default:
+			break;
+		}
+		trx_mutex_exit(trx);
+
+#endif /* WITH_WSREP */
```

Comment:
I see 2 redundant reads of `trx->error_state` in code paths where we already determined that `trx->error_state == DB_DEADLOCK`. Why are we acquring `trx->mutex` here?  A few lines later, we are again reading `trx->error_state` without mutex protection.

Where do you see that `trx_t::error_state` would be written by a thread that is not the one that is executing the transaction? There is a separate field `trx->lock.was_chosen_as_deadlock_victim` that can be set asynchronously.

Reply:
lock_cancel_waiting_release() may eventually change trx::error_state to DB_DEADLOCK, here is one call stack when external operator does such change:
innobase_kill_query
	lock_trx_handle_wait
		lock_trx_handle_wait_low
			lock_cancel_waiting_and_release
				lock_wait_release_thread_if_suspended
					trx->error_state = DB_DEADLOCK;
there are other scenarios as well.

The race condition, which this patch tries to fix happens, in row_ins_check_foreign_constraint(), where local variable 'err' has received DB_LOCK_WAIT code. In following call to lock_wait_suspend_thread(), it will clear trx->lock.was_chosen_as_deadlock_victim and set trx->error_state = DB_DEADLOCK. And, the old code used to keep  'err' with DB_LOCK_WAIT and return that, hence the problem.

Now, holding trx mutex may indeed be redundant here. External threads could only change error_state to DB_DEADLOCK, and it does not affect the logic in this fix. Only self can clear trx::error_state, and self is operating in this code path. 
But, torn read is only remaining vulnerability, high priority thread may write over this 4 byte enum, while victim trx is reading it here.

Note that this fix is only considering error state of DB_DEADLOCK. This is what galera slave applier is using for high priority aborts. During the testing, I noticed that error_state of DB_INTERRUPTED was also over written by DB_LOCK_WAIT. However, it does not affect this test execution.

Reply:
Thank you for the explanation.
I will let @Thirunarayanan confirm this and merge the fix to 10.2, since I am leaving for vacation. If a foreign thread can indeed assing to `trx->error_state` of other threads, then we should consider adding more `trx->mutex` protection. On my quick check of the code, I did not see evidence of that.